### PR TITLE
Feature/usage tracking, blueprint and pipeline builders, tests 

### PR DESCRIPTION
**Pull request by shapirov103**
*Issue #, if available:* #149
Fix for #152 (validated with version 0.11..4-test of the CDK blueprint)

*Description of changes:*
Feature/usage tracking, blueprint and pipeline builders, tests for usage and builders, CDK moved to 1.119 

The PR for patterns is coming (pipelines and builders). 

GitHub actions to add tests will also be added. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -23,6 +23,9 @@ list:
 	$(DEPS)
 	$(CDK) list
 
+run-test:
```
Comment (shapirov103) [inline Makefile:26]: Might be easier to follow if you start with test/stacks.test.ts: https://github.com/aws-quickstart/ssp-amazon-eks/pull/184/files#diff-59f0283006bf485e281b97535ff10c05afe6253b30c4ab3b6b298851ffd3579b

Comment (kcoleman731) [reply Makefile:26]: tests in this project are a wonderful thing. Thank you.

File: lib/pipelines/code-pipeline.ts [hunk 2]
```diff
@@ -2,57 +2,166 @@ import * as cdk from '@aws-cdk/core';
 import * as pipelines from '@aws-cdk/pipelines';
 import * as codepipeline from '@aws-cdk/aws-codepipeline';
 import * as actions from '@aws-cdk/aws-codepipeline-actions';
+import { Construct, StackProps } from '@aws-cdk/core';
```
Comment (kcoleman731) [inline lib/pipelines/code-pipeline.ts:5]: we should get aligned on when to spread imports and when to use `import * as`. We are inconsistent in the project currently.

File: lib/pipelines/code-pipeline.ts [hunk 3]
```diff
@@ -2,57 +2,166 @@ import * as cdk from '@aws-cdk/core';
...
+
+    private props: Partial<PipelineProps>;
 
-    public static build = (props: PipelineProps) => {
+
+    constructor() {
+        this.props = { stages: []};
+    }
+
+    public name(name: string): CodePipelineBuilder {
```
Comment (kcoleman731) [inline lib/pipelines/code-pipeline.ts:74]: are these meant to be setters? If so, consider renaming to `setName`. Also what does it return this?

File: lib/stacks/eks-blueprint-stack.ts [hunk 4]
```diff
@@ -38,23 +39,106 @@ export class EksBlueprintProps {
...
+        account?: string,
+        region?: string
+    }
+
+    constructor() {
+        this.props = { addOns: new Array<ClusterAddOn>(), teams: new Array<Team>()};
+        this.env = {};
+    }
+
+    public name(name: string) : this {
```
Comment (kcoleman731) [inline lib/stacks/eks-blueprint-stack.ts:69]: same thing here...this pattern seems odd to me. Also returning this? Is that valid typescript?

**Review by kcoleman731 — COMMENTED** [review]
@shapirov103 lets talk about where the builder pattern comes from? Is this a CDK pattern you have seen?

File: lib/pipelines/code-pipeline.ts (hunk 2 above)
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:5]: I mentioned elsewhere, I was using common conventions from my prev experience: wildcard after 5 imports.

File: lib/pipelines/code-pipeline.ts (hunk 3 above)
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:74]: They are not traditional setters. That the convention for the builder pattern. The difference is fluid API as each method returns `this`.  Alternative is adding `with` to each property but that seems overly verbose for little gain.

File: lib/stacks/eks-blueprint-stack.ts (hunk 4 above)
Comment (shapirov103) [reply lib/stacks/eks-blueprint-stack.ts:69]: I was using this for years with different programming languages. You can try `make build` `make run-test` `make lint` on the branch.

**Review by askulkarni2 — APPROVED** [review]
A couple of minor points but otherwise looks slick. Please kick off the e2e tests well when you are ready.

File: ci/buildspec.yml [hunk 5]
```diff
@@ -18,7 +18,6 @@ phases:
         make deps
       - >
         make bootstrap-cdk
-
   pre_build:
     commands:
       - >
```
[the remark this answers is not in the document]
Comment (shapirov103) [reply ci/buildspec.yml:23]: 100%

File: lib/pipelines/code-pipeline.ts [hunk 6]
```diff
@@ -65,13 +174,13 @@ export class CodePipeline {
                 // Install deps
                 'npm install',
                 // Install global CDK.
-                'npm install -g aws-cdk@1.113.0'
+                'npm install -g aws-cdk@1.119.0'
             ],
             buildCommands: ['npm run build'],
             synthCommand: 'cdk synth'
```
[the remark this answers is not in the document]
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:180]: Definitely. Will address.

File: lib/pipelines/code-pipeline.ts (hunk 3 above)
Comment (kcoleman731) [reply lib/pipelines/code-pipeline.ts:74]: gotcha. My vote would be to use setter function names. Makes it explicit that you are "setting" something as opposed to reading. Up to you.

File: lib/pipelines/code-pipeline.ts [hunk 7]
```diff
@@ -2,57 +2,166 @@ import * as cdk from '@aws-cdk/core';
...
+        return this;
+    }
+
+    public stage(stackStage: StackStage) : CodePipelineBuilder {
+        this.props.stages?.push(stackStage);
+        return this;
+    }
+    
+    build(scope: cdk.Construct, id: string, stackProps?: cdk.StackProps): cdk.Stack {
+        console.assert(this.props.name && this.props.owner && this.props.repository!.credentialsSecretName && this.props.stages,
```
Comment (kcoleman731) [inline lib/pipelines/code-pipeline.ts:95]: consider asserting on each individually. Helps be specific about which property is missing (and avoids a super long line of code!)

**Review by kcoleman731 — APPROVED** [review]

**Comment by kcoleman731** [comment]
@shapirov103 meta comment here...we should think about how to organize stacks and files in each stack. For instance, do we want to move the pipeline dir into stacks and then have `pipeline` directory and a `cluster` directory. Within the pipeline directory perhaps we have a stack.ts, a builder.ts, and a stage.ts. Same thing with the cluster directory...perhaps we have a stacks.ts and a builder.ts?

Not married to an approach but something to consider.

File: lib/pipelines/code-pipeline.ts (hunk 2 above)
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:5]: Appears to be resolved, as the number of imports before wildcards is 3. We need to look into making it a lint rule. So far I only observe a way to code that rule, so deferring for now.

File: lib/pipelines/code-pipeline.ts (hunk 3 above)
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:74]: Builders are used to reduce verbosity as one of the objectives and extra prefixes are not quite helping. I am leaning towards keeping it as is, even though i understand your perspective.

**Comment by shapirov103** [comment]
> @shapirov103 meta comment here...we should think about how to organize stacks and files in each stack. For instance, do we want to move the pipeline dir into stacks and then have `pipeline` directory and a `cluster` directory. Within the pipeline directory perhaps we have a stack.ts, a builder.ts, and a stage.ts. Same thing with the cluster directory...perhaps we have a stacks.ts and a builder.ts?
> 
> Not married to an approach but something to consider.

We can apply this, it is a bit more deterministic and consistent to what we have now. I will defer this one, probably a separate task or could be handled within your ES PR.

File: ci/buildspec.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (shapirov103) [reply ci/buildspec.yml:25]: Addressed

File: lib/pipelines/code-pipeline.ts (hunk 7 above)
Comment (shapirov103) [reply lib/pipelines/code-pipeline.ts:96]: Addressed

**Comment by shapirov103** [comment]
@kcoleman731 @askulkarni2 please take another look. 
Also the the PR to the patterns that was validated to work in my account (although against a branch). 
https://github.com/aws-samples/ssp-eks-patterns/pull/2

**Review by kcoleman731 — APPROVED** [review]

**Pull request merged**